File: src/main/java/io/github/saneea/dvh/feature/xml/FilteredComparison.java
package io.github.saneea.dvh.feature.xml;

import io.github.saneea.dvh.Feature;
import io.github.saneea.dvh.FeatureContext;
import org.apache.commons.cli.CommandLine;
import org.apache.commons.cli.Option;

import java.io.*;
import java.nio.charset.StandardCharsets;

public class FilteredComparison implements//
		Feature, //
		Feature.CLI, //
		Feature.CLI.Options {

	private CommandLine commandLine;

	@Override
	public Meta meta(FeatureContext context) {
		return Meta.from("run external comparison tool for 'pretty printed' xml files");
	}

	@Override
	public void run(FeatureContext context) throws Exception {
		String comparisonTool = commandLine.getOptionValue(Params.COMPARISON_TOOL);
		String leftFileName = commandLine.getOptionValue(Params.LEFT);
		String rightFileName = commandLine.getOptionValue(Params.RIGHT);

		String leftFilteredFileName;
		String rightFilteredFileName;
		try {
			leftFilteredFileName = getFilteredFile(leftFileName);
			rightFilteredFileName = getFilteredFile(rightFileName);
		} catch (Exception e) {
			leftFilteredFileName = leftFileName;
			rightFilteredFileName = rightFileName;
		}

		ProcessBuilder processBuilder = new ProcessBuilder(comparisonTool, leftFilteredFileName, rightFilteredFileName);
		Process process = processBuilder.start();
		process.waitFor();
	}

	private static String getFilteredFile(String inputFileName) throws Exception {
		File outputFile = File.createTempFile("filtered", null);
		outputFile.deleteOnExit();
		String outputFileName = outputFile.getPath();

		try (//
				Reader in = new InputStreamReader(//
						new BufferedInputStream(//
								new FileInputStream(//
										inputFileName)), //
						StandardCharsets.UTF_8); //

				Writer out = new OutputStreamWriter(//
						new BufferedOutputStream(//
								new FileOutputStream(//
										outputFile)), //
						StandardCharsets.UTF_8);//
		) {
			XmlPrettyPrint.Companion.run(in, out, false);
		}
		return outputFileName;
	}

	public static class Params {

		public static String COMPARISON_TOOL = "comparisonTool";
		public static String LEFT = "left";
		public static String RIGHT = "right";

		private static Option[] createOptions() {
			Option[] options = {

					Option//
							.builder("ct")//
							.longOpt(COMPARISON_TOOL)//
							.hasArg(true)//
							.argName("system command")//
							.required(true)//
							.desc("system command for comparison of just path to extrenal comparison tool")//
							.build(), //
					Option//
							.builder("l")//
							.longOpt(LEFT)//
							.hasArg(true)//
							.argName("file path")//
							.required(true)//
							.desc("left side file for comparison")//
							.build(), //
					Option//
							.builder("r")//
							.longOpt(RIGHT)//
							.hasArg(true)//
							.argName("file path")//
							.required(true)//
							.desc("right side file for comparison")//
							.build()//
			};

			return options;
		}
	}

	@Override
	public void setCommandLine(CommandLine commandLine) {
		this.commandLine = commandLine;
	}

	@Override
	public Option[] getOptions() {
		return Params.createOptions();
	}

}

File: src/main/kotlin/io/github/saneea/dvh/feature/xml/FilteredComparison.kt
package io.github.saneea.dvh.feature.xml

import io.github.saneea.dvh.Feature
import io.github.saneea.dvh.Feature.Meta
import io.github.saneea.dvh.feature.xml.XmlPrettyPrint.Companion.run
import org.apache.commons.cli.CommandLine
import org.apache.commons.cli.Option
import java.io.*
import java.nio.charset.StandardCharsets

class FilteredComparison :
    Feature,
    Feature.CLI,
    Feature.CLI.Options {

    override lateinit var commandLine: CommandLine

    override val meta = Meta("run external comparison tool for 'pretty printed' xml files")

    override fun run() {
        val comparisonTool = commandLine.getOptionValue(Params.COMPARISON_TOOL)
        val leftFileName = commandLine.getOptionValue(Params.LEFT)
        val rightFileName = commandLine.getOptionValue(Params.RIGHT)

        var leftFilteredFileName: String
        var rightFilteredFileName: String
        try {
            leftFilteredFileName = getFilteredFile(leftFileName)
            rightFilteredFileName = getFilteredFile(rightFileName)
        } catch (e: Exception) {
            leftFilteredFileName = leftFileName
            rightFilteredFileName = rightFileName
        }

        val processBuilder = ProcessBuilder(comparisonTool, leftFilteredFileName, rightFilteredFileName)
        val process = processBuilder.start()
        process.waitFor()
    }

    object Params {
        var COMPARISON_TOOL = "comparisonTool"
        var LEFT = "left"
        var RIGHT = "right"

        fun createOptions(): Array<Option> {
            return arrayOf(
                Option
                    .builder("ct")
                    .longOpt(COMPARISON_TOOL)
                    .hasArg(true)
                    .argName("system command")
                    .required(true)
                    .desc("system command for comparison of just path to extrenal comparison tool")
                    .build(),
                Option
                    .builder("l")
                    .longOpt(LEFT)
                    .hasArg(true)
                    .argName("file path")
                    .required(true)
                    .desc("left side file for comparison")
                    .build(),
                Option
                    .builder("r")
                    .longOpt(RIGHT)
                    .hasArg(true)
                    .argName("file path")
                    .required(true)
                    .desc("right side file for comparison")
                    .build()
            )
        }
    }

    override val options get() = Params.createOptions()

    private fun getFilteredFile(inputFileName: String): String {
        val outputFile = File.createTempFile("filtered", null)
        outputFile.deleteOnExit()
        val outputFileName = outputFile.path
        InputStreamReader(
            BufferedInputStream(
                FileInputStream(
                    inputFileName
                )
            ),
            StandardCharsets.UTF_8
        ).use { `in` ->
            OutputStreamWriter(
                BufferedOutputStream(
                    FileOutputStream(
                        outputFile
                    )
                ),
                StandardCharsets.UTF_8
            ).use { out -> run(`in`, out, false) }
        }
        return outputFileName
    }

}